File: sdk/node/mcp-auth-middleware.js
/**
 * Universal MCP Authentication Middleware (Node.js)
 * 
 * DESIGN PRINCIPLE:
 * This middleware is 100% LLM-agnostic. It does not know or care whether
 * Claude Desktop, Cursor, ChatGPT, Gemini, or a custom agent is calling.
 * 
 * Works identically for:
 * - Mode 1: Dynamic OAuth 2.1 tokens
 * - Mode 2: Static long-lived API key tokens
 * 
 * ZERO EXTERNAL DEPENDENCIES:
 * Built purely on Node.js standard modules (`crypto`, `http`, `https`).
 */

const crypto = require('crypto');
const http = require('http');
const https = require('https');
const { URL } = require('url');

/**
 * Creates the MCP Authentication Middleware for Express / Connect / HTTP servers.
 *
 * @param {Object} options
 * @param {string} options.jwksUri - Full URL to central auth server JWKS endpoint (e.g. 'http://localhost:3000/.well-known/jwks.json')
 * @param {string} options.audience - Expected audience for this MCP server (e.g. 'mcp-invoicing')
 * @param {string} [options.revocationsUri] - Optional URL to revocation list endpoint (e.g. 'http://localhost:3000/revocations')
 * @param {number} [options.jwksCacheTtlMs=300000] - JWKS cache TTL (default 5 minutes)
 * @param {number} [options.revocationCacheTtlMs=30000] - Revocation list cache TTL (default 30 seconds)
 * @param {number} [options.clockSkewSeconds=60] - Allowed clock skew for expiration
 */
function createMcpAuthMiddleware(options) {
  if (!options || !options.jwksUri || !options.audience) {
    throw new Error('[McpAuth] Both `jwksUri` and `audience` options are required.');
  }

  const {
    jwksUri,
    audience,
    revocationsUri = options.jwksUri.replace('/.well-known/jwks.json', '/revocations'),
    jwksCacheTtlMs = 5 * 60 * 1000,
    revocationCacheTtlMs = 30 * 1000,
    clockSkewSeconds = 60
  } = options;

  let keyCache = {
    keys: new Map(),
    expiresAt: 0
  };

  let revocationCache = {
    revokedClientIds: new Set(),
    revokedJtis: new Set(),
    expiresAt: 0
  };

  // HTTP GET JSON Helper (Standard Library only)
  function fetchJson(urlStr) {
    return new Promise((resolve, reject) => {
      const url = new URL(urlStr);
      const client = url.protocol === 'https:' ? https : http;

      const req = client.get(url, (res) => {
        if (res.statusCode < 200 || res.statusCode >= 300) {
          return reject(new Error(`HTTP ${res.statusCode} fetching ${urlStr}`));
        }
        let raw = '';
        res.on('data', (chunk) => (raw += chunk));
        res.on('end', () => {
          try {
            resolve(JSON.parse(raw));
          } catch (e) {
            reject(new Error(`Failed to parse JSON response from ${urlStr}`));
          }
        });
      });

      req.on('error', reject);
      req.setTimeout(5000, () => {
        req.destroy();
        reject(new Error(`Timeout fetching ${urlStr}`));
      });
    });
  }

  // Refresh JWKS Public Key Cache
  async function getPublicKey(kid) {
    const now = Date.now();
    if (now < keyCache.expiresAt && keyCache.keys.has(kid)) {
      return keyCache.keys.get(kid);
    }

    try {
      const jwks = await fetchJson(jwksUri);
      if (!jwks || !Array.isArray(jwks.keys)) {
        throw new Error('Invalid JWKS payload');
      }

      keyCache.keys.clear();
      for (const jwk of jwks.keys) {
        if (jwk.kty === 'RSA') {
          const keyObject = crypto.createPublicKey({ key: jwk, format: 'jwk' });
          keyCache.keys.set(jwk.kid, keyObject);
        }
      }
      keyCache.expiresAt = now + jwksCacheTtlMs;

      if (kid && keyCache.keys.has(kid)) {
        return keyCache.keys.get(kid);
      }
      // If no kid match, return first key as fallback if single key
      if (keyCache.keys.size === 1) {
        return keyCache.keys.values().next().value;
      }
      return null;
    } catch (err) {
      console.error('[McpAuth] Failed to refresh JWKS:', err.message);
      return keyCache.keys.get(kid) || null;
    }
  }

  // Refresh Revocation Cache
  async function refreshRevocations() {
    if (!revocationsUri) return;
    const now = Date.now();

    if (now >= revocationCache.expiresAt) {
      try {
        const data = await fetchJson(revocationsUri);
        if (data) {
          if (Array.isArray(data.revoked_client_ids)) {
            revocationCache.revokedClientIds = new Set(data.revoked_client_ids);
          }
          if (Array.isArray(data.revoked_jtis)) {
            revocationCache.revokedJtis = new Set(data.revoked_jtis);
          }
          revocationCache.expiresAt = now + revocationCacheTtlMs;
        }
      } catch (err) {
        // Soft fail on network issue, preserve existing cache
      }
    }
  }

  async function isClientRevoked(clientId) {
    if (!clientId) return false;
    await refreshRevocations();
    return revocationCache.revokedClientIds.has(clientId);
  }

  async function isJtiRevoked(jti) {
    if (!jti) return false;
    await refreshRevocations();
    return revocationCache.revokedJtis.has(jti);
  }

  // Pure Node JWT Verification
  async function verifyJwt(token) {
    const parts = token.split('.');
    if (parts.length !== 3) {
      throw new Error('Malformed token structure');
    }

    const [headerB64, payloadB64, signatureB64] = parts;

    // Decode Header & Payload
    let header, payload;
    try {
      header = JSON.parse(Buffer.from(headerB64, 'base64url').toString('utf8'));
      payload = JSON.parse(Buffer.from(payloadB64, 'base64url').toString('utf8'));
    } catch (e) {
      throw new Error('Invalid token JSON encoding');
    }

    if (header.alg !== 'RS256') {
      throw new Error(`Unsupported signature algorithm: ${header.alg}. Only RS256 is accepted.`);
    }

    // Retrieve public key from JWKS
    const publicKey = await getPublicKey(header.kid);
    if (!publicKey) {
      throw new Error(`Public key not found in JWKS for kid: ${header.kid}`);
    }

    // Verify RS256 signature
    const verifier = crypto.createVerify('RSA-SHA256');
    verifier.update(`${headerB64}.${payloadB64}`);
    const signature = Buffer.from(signatureB64, 'base64url');

    const isValid = verifier.verify(publicKey, signature);
    if (!isValid) {
      throw new Error('Invalid token cryptographic signature');
    }

    // Verify Audience
    const aud = payload.aud;
    const audMatches = Array.isArray(aud) ? aud.includes(audience) : aud === audience;
    if (!audMatches) {
      throw new Error(`Audience mismatch. Token aud: "${aud}", expected: "${audience}"`);
    }

    // Verify Expiration
    const nowSeconds = Math.floor(Date.now() / 1000);
    if (payload.exp && (nowSeconds - clockSkewSeconds) > payload.exp) {
      throw new Error(`Token expired at ${new Date(payload.exp * 1000).toISOString()}`);
    }

    // Verify Revocation (Client-level and JTI-level)
    const clientId = payload.client_id || payload.sub;
    if (await isClientRevoked(clientId)) {
      throw new Error(`Client "${clientId}" has been revoked by the administrator`);
    }

    if (payload.jti && await isJtiRevoked(payload.jti)) {
      throw new Error('Token has been revoked or rotated');
    }

    return payload;
  }

  // Return Express / Connect Middleware
  return async function mcpAuthMiddleware(req, res, next) {
    let token = null;

    // 1. Check standard Authorization: Bearer <token>
    const authHeader = req.headers.authorization || req.headers.Authorization;
    if (authHeader) {
      const parts = authHeader.split(' ');
      if (parts.length === 2 && parts[0].toLowerCase() === 'bearer') {
        token = parts[1].trim();
      }
    }

    // 2. Fallback: Check static x-api-key / x-mcp-api-key headers (Mode 2)
    if (!token) {
      token = req.headers['x-api-key'] || req.headers['x-mcp-api-key'];
    }

    // 3. Fallback: URL query parameter for SSE or Stream transports
    if (!token && req.query && req.query.access_token) {
      token = req.query.access_token;
    }

    if (!token) {
      res.setHeader('WWW-Authenticate', `Bearer realm="${audience}", error="unauthorized"`);
      return res.status(401).json({
        error: 'unauthorized',
        error_description: 'Missing authentication token. Provide Bearer token or x-api-key header.'
      });
    }

    try {
      const payload = await verifyJwt(token);
      // Attach verified auth context to request object
      req.auth = payload;
      req.mcpClient = {
        clientId: payload.client_id || payload.sub,
        audience: payload.aud,
        authMode: payload.auth_mode,
        jti: payload.jti
      };
      next();
    } catch (err) {
      res.setHeader('WWW-Authenticate', `Bearer realm="${audience}", error="invalid_token", error_description="${err.message}"`);
      return res.status(401).json({
        error: 'invalid_token',
        error_description: err.message
      });
    }
  };
}

module.exports = {
  createMcpAuthMiddleware
};

File: sdk/python/mcp_auth_middleware.py
"""
Universal MCP Authentication Middleware (Python / FastAPI / Starlette)

DESIGN PRINCIPLE:
This middleware is 100% LLM-agnostic. It does not know or care whether
Claude Desktop, Cursor, ChatGPT, Gemini, or a custom agent is calling.

Works identically for:
- Mode 1: Dynamic OAuth 2.1 tokens
- Mode 2: Static long-lived API key tokens

REQUIREMENTS:
pip install PyJWT[crypto] requests
"""

import time
import requests
import jwt
from jwt import PyJWKClient
from typing import Optional, Set
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import JSONResponse

class McpAuthMiddleware(BaseHTTPMiddleware):
    def __init__(
        self,
        app,
        jwks_uri: str,
        audience: str,
        revocations_uri: Optional[str] = None,
        jwks_cache_ttl: int = 300,
        revocation_cache_ttl: int = 30,
        clock_skew_seconds: int = 60,
        exempt_paths: Optional[list] = None
    ):
        super().__init__(app)
        self.jwks_uri = jwks_uri
        self.audience = audience
        self.revocations_uri = revocations_uri or jwks_uri.replace('/.well-known/jwks.json', '/revocations')
        self.clock_skew_seconds = clock_skew_seconds
        self.exempt_paths = set(exempt_paths) if exempt_paths else {"/health", "/healthz"}

        # PyJWKClient handles JWKS retrieval and caching
        self.jwks_client = PyJWKClient(jwks_uri, cache_keys=True, max_cached_keys=16)

        # Revocation cache
        self.revocation_cache_ttl = revocation_cache_ttl
        self.revocation_cache_expires_at = 0.0
        self.revoked_client_ids: Set[str] = set()
        self.revoked_jtis: Set[str] = set()

    def _refresh_revocations(self):
        now = time.time()
        if now >= self.revocation_cache_expires_at:
            try:
                resp = requests.get(self.revocations_uri, timeout=5)
                if resp.status_code == 200:
                    data = resp.json()
                    self.revoked_client_ids = set(data.get("revoked_client_ids", []))
                    self.revoked_jtis = set(data.get("revoked_jtis", []))
                    self.revocation_cache_expires_at = now + self.revocation_cache_ttl
            except Exception as e:
                # Retain old cache on transient network error
                pass

    def _is_revoked(self, client_id: str) -> bool:
        if not client_id:
            return False
        self._refresh_revocations()
        return client_id in self.revoked_client_ids

    def _is_jti_revoked(self, jti: str) -> bool:
        if not jti:
            return False
        self._refresh_revocations()
        return jti in self.revoked_jtis

    async def dispatch(self, request: Request, call_next):
        path = request.url.path
        if any(path == ep or path.startswith(ep.rstrip("/") + "/") for ep in self.exempt_paths):
            return await call_next(request)

        token = None

        # 1. Check standard Authorization: Bearer <token>
        auth_header = request.headers.get("Authorization") or request.headers.get("authorization")
        if auth_header:
            parts = auth_header.split()
            if len(parts) == 2 and parts[0].lower() == "bearer":
                token = parts[1].strip()

        # 2. Check static x-api-key header (Mode 2)
        if not token:
            token = request.headers.get("x-api-key") or request.headers.get("x-mcp-api-key")

        # 3. Check query param (for SSE / Stream transport)
        if not token and "access_token" in request.query_params:
            token = request.query_params["access_token"]

        if not token:
            return JSONResponse(
                status_code=401,
                content={
                    "error": "unauthorized",
                    "error_description": "Missing authentication token. Provide Bearer or x-api-key header."
                },
                headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="unauthorized"'}
            )

        try:
            # 1. Fetch matching signing key from JWKS
            signing_key = self.jwks_client.get_signing_key_from_jwt(token)

            # 2. Verify RS256 signature, audience, and expiration
            payload = jwt.decode(
                token,
                signing_key.key,
                algorithms=["RS256"],
                audience=self.audience,
                leeway=self.clock_skew_seconds
            )

            # 3. Check revocation status (client-level and individual token JTI)
            client_id = payload.get("client_id") or payload.get("sub")
            if self._is_revoked(client_id):
                return JSONResponse(
                    status_code=401,
                    content={
                        "error": "invalid_token",
                        "error_description": f'Client "{client_id}" has been revoked.'
                    },
                    headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="invalid_token"'}
                )

            jti = payload.get("jti")
            if self._is_jti_revoked(jti):
                return JSONResponse(
                    status_code=401,
                    content={
                        "error": "invalid_token",
                        "error_description": "Token has been revoked or rotated."
                    },
                    headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="invalid_token"'}
                )

            # Attach auth state to request
            request.state.auth = payload
            request.state.mcp_client_id = client_id

            return await call_next(request)

        except jwt.ExpiredSignatureError:
            return JSONResponse(
                status_code=401,
                content={"error": "invalid_token", "error_description": "Token has expired."},
                headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="invalid_token"'}
            )
        except jwt.InvalidAudienceError:
            return JSONResponse(
                status_code=401,
                content={"error": "invalid_token", "error_description": f'Audience mismatch. Expected "{self.audience}".'},
                headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="invalid_token"'}
            )
        except Exception as e:
            return JSONResponse(
                status_code=401,
                content={"error": "invalid_token", "error_description": str(e)},
                headers={"WWW-Authenticate": f'Bearer realm="{self.audience}", error="invalid_token"'}
            )
